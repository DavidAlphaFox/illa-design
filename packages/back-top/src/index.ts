export * from "./back-top"
export * from "./interface"
export * from "./style"
