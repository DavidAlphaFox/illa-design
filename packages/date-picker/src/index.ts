export * from "./date-picker"
export * from "./interface"
export * from "./style"
