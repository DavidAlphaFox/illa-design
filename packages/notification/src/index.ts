export * from "./interface"
export * from "./notification"
export * from "./style"
