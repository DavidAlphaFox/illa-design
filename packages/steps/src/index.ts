export * from "./steps"
export * from "./step"
export * from "./interface"
