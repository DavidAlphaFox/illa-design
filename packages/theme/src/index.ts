export * from "./global-color"
export * from "./interface"
export * from "./color-palette"
export * from "./style"
