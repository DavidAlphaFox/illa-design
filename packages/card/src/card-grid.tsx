import { Children, cloneElement, forwardRef, ReactElement } from "react"
import { CardGridProps } from "./interface"
import { applyCardGrid } from "./style"
import { applyBoxStyle } from "@illa-design/theme"
import { deleteCssProps } from "@illa-design/theme"

export const CardGrid = forwardRef<HTMLDivElement, CardGridProps>(
  (props, ref) => {
    const { hoverable, children, ...restProps } = props
    const renderChildren = Children.map(
      children as ReactElement,
      (element: JSX.Element) => {
        if (element && element.type && element.type.displayName === "Card") {
          return cloneElement(element, { isGridMode: true })
        }
        return element
      },
    )
    return (
      <div
        ref={ref}
        css={[applyCardGrid(hoverable ?? false), applyBoxStyle(props)]}
        {...deleteCssProps(restProps)}
      >
        {renderChildren}
      </div>
    )
  },
)

CardGrid.displayName = "CardGrid"
