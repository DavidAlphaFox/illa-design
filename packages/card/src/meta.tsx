import { forwardRef } from "react"
import { CardMetaProps } from "./interface"
import {
  applyCardMetaDescription,
  applyCardMetaFooter,
  applyCardMetaTitle,
} from "./style"
import { applyBoxStyle } from "@illa-design/theme"
import { deleteCssProps } from "@illa-design/theme"

export const Meta = forwardRef<HTMLDivElement, CardMetaProps>((props, ref) => {
  const { title, description, actionList, avatar, ...restProps } = props
  return (
    <div ref={ref} css={applyBoxStyle(props)} {...deleteCssProps(restProps)}>
      {title || description ? (
        <div>
          {title && <div css={applyCardMetaTitle}>{title}</div>}
          {description && (
            <div css={applyCardMetaDescription}>{description}</div>
          )}
        </div>
      ) : null}
      {avatar || actionList ? (
        <div css={applyCardMetaFooter}>
          {avatar ? <div>{avatar}</div> : null}
          {actionList}
        </div>
      ) : null}
    </div>
  )
})

Meta.displayName = "Meta"
