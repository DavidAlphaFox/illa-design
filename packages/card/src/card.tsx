import { forwardRef, Children, cloneElement, ReactElement } from "react"
import { CardGrid } from "./card-grid"
import { Meta } from "./meta"
import { CardProps } from "./interface"
import { Loading } from "@illa-design/loading"
import {
  applyCardActionItem,
  applyCard,
  applyCardActions,
  applyCardBody,
  applyCardCover,
  applyCardHeader,
  applyCardHeaderExtra,
  applyCardHeaderTitle,
  applyCardActionsRight,
} from "./style"
import { applyBoxStyle } from "@illa-design/theme"

export const Card = forwardRef<HTMLDivElement, CardProps>((props, ref) => {
  const {
    title,
    actions,
    extra,
    size = "medium",
    hoverable,
    cover,
    loading,
    bordered = true,
    children,
    isGridMode,
    ...restProps
  } = props

  let isContainGrid = false
  let isContainMeta = false

  Children.forEach(children, (element) => {
    if (element && (element as ReactElement).type) {
      if ((element as ReactElement).type === CardGrid) {
        isContainGrid = true
      } else if ((element as ReactElement).type === Meta) {
        isContainMeta = true
      }
    }
  })

  const actionList =
    actions && actions.length ? (
      <div css={applyCardActions(isContainMeta)}>
        <div css={applyCardActionsRight}>
          {actions.map((action, index) => (
            <span key={`action-${index}`} css={applyCardActionItem}>
              {action}
            </span>
          ))}
        </div>
      </div>
    ) : null

  const handledChildren = Children.map(
    children as ReactElement,
    (element: JSX.Element) => {
      if (element && element.type && element.type === Meta) {
        return cloneElement(element, { actionList })
      }
      return element
    },
  )

  return (
    <div
      ref={ref}
      css={[
        applyCard(bordered, hoverable ?? false, isGridMode),
        applyBoxStyle(props),
      ]}
      {...restProps}
    >
      {title || extra ? (
        <div css={applyCardHeader(size)}>
          {title && <div css={applyCardHeaderTitle}>{title}</div>}
          {extra && <div css={applyCardHeaderExtra}>{extra}</div>}
        </div>
      ) : null}

      {cover ? <div css={applyCardCover}>{cover}</div> : null}

      <div css={applyCardBody(size, loading ?? false, isContainGrid)}>
        {loading ? <Loading size={size} colorScheme="blue" /> : handledChildren}
        {isContainMeta ? null : actionList}
      </div>
    </div>
  )
})

Card.displayName = "Card"
