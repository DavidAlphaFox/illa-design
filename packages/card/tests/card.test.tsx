import { Card, CardMeta } from "../src"
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import { Avatar } from "@illa-design/avatar"
import { LikeIcon, MoreIcon, ShareIcon } from "@illa-design/icon"
import "@testing-library/jest-dom"
import { Link } from "@illa-design/link"

test("Card renders with different size", () => {
  render(
    <Card size="small" title="small" extra={<Link>More</Link>}>
      <p>Hello</p>
      <p>Hello</p>
      <p>Hello</p>
    </Card>,
  )
  render(<Card title="medium" />)
  expect(screen.getByText("small")).toHaveStyle({
    padding: 8,
  })
  expect(screen.getByText("medium")).toHaveStyle({
    padding: 16,
  })
})

test("Card renders with different hoverable", () => {
  render(<Card hoverable title="hoverable" />)
  userEvent.hover(screen.getByText("hoverable"))
  expect(screen.getByText("hoverable").parentNode?.parentNode).toMatchSnapshot()
})

test("Card renders with Meta", () => {
  render(
    <CardMeta
      title="CardMeta"
      description="MetaContent"
      avatar={<Avatar />}
      actionList={[
        <LikeIcon key="like" />,
        <ShareIcon key="share" />,
        <MoreIcon key="more" />,
      ]}
    ></CardMeta>,
  )
  expect(screen.getByText("CardMeta").parentNode?.parentNode).toMatchSnapshot()
  expect(screen.getByText("CardMeta").parentNode?.nextSibling).toHaveStyle({
    marginTop: 16,
  })
})

test("Card renders with Cover", () => {
  render(
    <CardMeta
      title="Cover"
      cover={
        <div
          style={{
            height: 204,
            overflow: "hidden",
          }}
        >
          <img
            style={{ width: "100%", transform: "translateY(-20px)" }}
            alt="dessert"
            src="https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png"
          />
        </div>
      }
    />,
  )
  expect(screen.getByText("Cover").parentNode?.nextSibling).toMatchSnapshot()
})
