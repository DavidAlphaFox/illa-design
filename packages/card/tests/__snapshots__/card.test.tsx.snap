// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders with Cover 1`] = `null`;

exports[`Card renders with Meta 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: box-shadow 0.2s ease-in-out;
  transition: box-shadow 0.2s ease-in-out;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  padding: 16px;
  background-color: #ffffffff;
  border-radius: 8px;
  color: #a9aeb8ff;
  border: solid 1px #e5e6ebff;
}

.emotion-2 {
  font-size: 16px;
  font-weight: 500;
  color: #1d2129ff;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-3 {
  margin-top: 4px;
  color: #a9aeb8ff;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 16px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-6 {
  margin-right: 8px;
}

.emotion-7 {
  vertical-align: middle;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-8 {
  background-color: #c2c2c2ff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #ffffffff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 32px;
  height: 32px;
  font-size: 14px;
  line-height: 22px;
  border-radius: 50%;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 14px;
  font-size: 14px;
  height: 14px;
}

.emotion-10 {
  vertical-align: middle;
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-right: 24px;
  margin-bottom: 0px;
}

.emotion-17 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-right: 0px;
  margin-bottom: 0px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        CardMeta
      </div>
      <div
        class="emotion-3"
      >
        MetaContent
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        >
          <div
            class="emotion-6"
          >
            <div
              class="emotion-7"
            >
              <div
                class="emotion-8"
              >
                <div
                  class="emotion-9"
                >
                  <svg
                    class="emotion-10"
                    color="currentColor"
                    fill="none"
                    height="14px"
                    viewBox="0 0 16 16"
                    width="14px"
                  >
                    <title>
                      PersonIcon
                    </title>
                    <path
                      d="M8 8a4 4 0 1 0 0-8 4 4 0 0 0 0 8zm2.667-4a2.667 2.667 0 1 1-5.334 0 2.667 2.667 0 0 1 5.334 0zM16 14.667C16 16 14.667 16 14.667 16H1.333S0 16 0 14.667c0-1.334 1.333-5.334 8-5.334s8 4 8 5.334zm-1.333-.006c-.002-.328-.206-1.314-1.11-2.218-.869-.87-2.505-1.776-5.557-1.776-3.053 0-4.688.906-5.557 1.776-.904.904-1.107 1.89-1.11 2.218h13.334z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="emotion-11"
        >
          <div
            class="emotion-12"
          >
            <div
              class="emotion-13"
            >
              <svg
                class="emotion-10"
                color="currentColor"
                fill="none"
                height="1em"
                viewBox="0 0 16 16"
                width="1em"
              >
                <title>
                  LikeIcon
                </title>
                <path
                  clip-rule="evenodd"
                  d="m8.892.83.121.057c.811.391 1.988 1.18 1.988 2.447 0 .502-.223 1.168-.669 1.998h2.526c1.99 0 2.447 1.756 1.99 3.071l-1.99 5.237a1.414 1.414 0 0 1-1.366 1.032H3.328v-9.34h1.365l2.967-4.22C7.847.812 8.353.584 8.892.83zM2.333 5.333v9.333H1V5.333h1.333z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="emotion-13"
            >
              <svg
                class="emotion-10"
                color="currentColor"
                fill="none"
                height="1em"
                viewBox="0 0 16 16"
                width="1em"
              >
                <title>
                  ShareIcon
                </title>
                <path
                  clip-rule="evenodd"
                  d="M8.585 4.644C3.844 4.644 0 8.53 0 13.324c0 .297.032.697.097 1.2a.195.195 0 0 0 .359.079l.123-.193c.12-.182.223-.33.312-.444 1.4-1.795 3.687-2.81 6.7-2.85l.994-.002v3.49a.404.404 0 0 0 .685.28l6.554-6.464a.588.588 0 0 0 0-.84L9.27 1.116a.405.405 0 0 0-.568 0 .393.393 0 0 0-.118.28v3.248z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="emotion-17"
            >
              <svg
                class="emotion-10"
                color="currentColor"
                fill="none"
                height="1em"
                viewBox="0 0 12 12"
                width="1em"
              >
                <title>
                  MoreIcon
                </title>
                <path
                  clip-rule="evenodd"
                  d="M2 6a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm5 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm4 1a1 1 0 1 0 0-2 1 1 0 0 0 0 2z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Card renders with different hoverable 1`] = `
.emotion-0 {
  font-size: 14px;
  background-color: #ffffffff;
  border-radius: 8px;
  -webkit-transition: box-shadow 0.2s ease-in-out;
  transition: box-shadow 0.2s ease-in-out;
  border: solid 1px #e5e6ebff;
}

.emotion-0:hover {
  box-shadow: 0 4px 10px 0 #00000013;
}

.emotion-1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  border-bottom: solid 1px #e5e6ebff;
  padding: 16px;
}

.emotion-2 {
  font-size: 16px;
  font-weight: 500;
  line-height: 1.5;
  color: #1d2129ff;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-3 {
  box-sizing: border-box;
  padding: 16px;
  color: #a9aeb8ff;
}

<div
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    >
      hoverable
    </div>
  </div>
  <div
    class="emotion-3"
  />
</div>
`;
